WordPress/gutenberg
# Introduce a dedicated autosaves endpoint for handling autosave behavior

## Description

Based on previous work from https://github.com/WordPress/gutenberg/pull/4156.

Leverages the final version of the Autosaves controller from @azaozz in Trac, https://core.trac.wordpress.org/ticket/43316#comment:76

See https://github.com/WordPress/gutenberg/issues/4112.

Does not implement the autosave alert when an autosave exists and the deletion of stale autosaves on load, these are handled in https://github.com/WordPress/gutenberg/pull/4218

## How Has This Been Tested?
Local testing, db monitoring. Verified behavior as expected, matches core.

* Create a new post, don't type anything
* Check the database, there will be a single auto-draft
* type a title and some content, pause for 10 seconds
* note the message 'Autosaving' ... 'Saved' ... 'Save draft'
* check the database, not that there is a single row for the post, with a draft status
* Click the 'save draft' button
* note the message 'Saving' ... 'Saved'  - button stays in saved state
* check the database, note a revision has been created because of the manual save action (not an autosave)
* publish the post
* check the database, note the post status is now published
* make some more changes and wait 10 seconds
* note the message 'Autosaving' ... then disappears (not save draft option) - @todo maybe should show saved briefly here?
* check the database, note an autosave revision has been created
* make some more changes and wait again for the autosave
* check the database and note the autosave is updated, and that there is only one autosave
* note that the 'update button' is available, click it
* note that 'update' is disabled (until you change the post)
* check the database, note that a new revision was created
* repeat the last three steps
* exit the post edit screen and log in as a different user, open the same post
* edit the post and wait for an autosave
* check the database, note that there are now two autosaves for the post, one for each user.

## Types of changes
* New WP_REST_Autosaves_Controller
* New REQUEST_POST_AUTOSAVE effect, autosavePost action
* Tracks `isPostAutosavable` separately from `isEditedPostSaveable`
* Leverage `withChangeDetection` to track autosave state (this needs more testing) 

## Checklist:
